Frameworks/ImageIO/CGImageDestination.mm

```diff
@@ -28,75 +885,358 @@ CGImageDestinationRef CGImageDestinationCreateWithDataConsumer(CGDataConsumerRef
+                loopCountMSB = (loopCount >> 8) & 0xff;
+            }
+        }
+
+        ComPtr<IWICMetadataQueryWriter> imageMetadataQueryWriter = imageDestination.idGifEncoderMetadataQueryWriter;
+        PROPVARIANT propertyToWrite;
+
+        PropVariantInit(&propertyToWrite);
+        propertyToWrite.vt = VT_VECTOR | VT_UI1;
+        propertyToWrite.caub.cElems = 11;
+        propertyToWrite.caub.pElems = (unsigned char*)[@"NETSCAPE2.0" UTF8String];
+        writePropertyToFrame(&propertyToWrite, L"/appext/Application", imageMetadataQueryWriter.Get());
+
+        propertyToWrite.caub.cElems = 5;
+        propertyToWrite.caub.pElems =
+            (unsigned char*)[[NSString stringWithFormat:@"%c%c%c%c%c", 3, 1, loopCountLSB, loopCountMSB, 0] UTF8String];
+        writePropertyToFrame(&propertyToWrite, L"/appext/Data", imageMetadataQueryWriter.Get());
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Current release supports JPEG, BMP, PNG, GIF, & TIFF image formats only 
 */
 bool CGImageDestinationFinalize(CGImageDestinationRef idst) {
-    UNIMPLEMENTED();
-    return StubReturn();
-}
+    if (!idst) {
+        return false;
+    }
+
+    ImageDestination* imageDestination = (ImageDestination*)idst;
+
+    if (imageDestination.count != imageDestination.maxCount) {
+        NSTraceInfo(TAG, @"CGImageDestinationFinalize image destination does not have enough images\n");
+        return false;
+    }
+
+    ComPtr<IWICBitmapEncoder> imageEncoder = imageDestination.idEncoder;
+    ComPtr<IWICStream> imageStream = imageDestination.idStream;
+
+    // If Encoder or Stream are missing, the destination has either been already finalized or not initialized yet, so return
+    if (!imageEncoder || !imageStream) {
+        NSTraceInfo(TAG, @"CGImageDestinationFinalize did not find an Encoder or Stream\n");
+        return false;
+    }
+
+    HRESULT status = imageEncoder->Commit();
+    if (!SUCCEEDED(status)) {
+        NSTraceInfo(TAG, @"Encoder Commit failed with status=%x\n", status);
+        return false;
+    }
+
+    if (imageDestination.outData) {
+        NSMutableData* dataNSPointer = static_cast<NSMutableData*>(imageDestination.outData);
+
+        // Seek to beginning of stream after image data all written
+        status = imageStream->Seek({0}, STREAM_SEEK_SET, NULL);
+        if (!SUCCEEDED(status)) {
+            NSTraceInfo(TAG, @"Stream Seek failed with status=%x\n", status);
+            return false;
+        }
+
+        // Get stream stats in order to determine number of bytes that were written
+        STATSTG streamStats;
+        status = imageStream->Stat(&streamStats, STATFLAG_NONAME);
+        if (!SUCCEEDED(status)) {
+            NSTraceInfo(TAG, @"Fetch stream stats failed with status=%x\n", status);
+            return false;
+        }
+
+        // Copy stream into the mutable data
+        [dataNSPointer increaseLengthBy:streamStats.cbSize.QuadPart];
```

Comment:
just want to confirm that this is only the size of the metadata that we written; we're not allocating more than we need to here, right?